docs/CBOR.md

```diff
@@ -0,0 +1,139 @@
+
+## Basic usage with `GenCodec`
+
+Nothing else than a `GenCodec` instance is necessary to serialize a type into CBOR. For example:
+
+```scala
+import com.avsystem.commons.serialization._
+import com.avsystem.commons.serialization.cbor._
+
+case class Thing(int: Int, str: String)
+object Thing extends HasGenCodec[Thing]
+
+val cborBytes: Array[Byte] = CborOutput.write(Thing(42, "foo"))
+```
+
+### `RawCbor`
+
+You can also serialize into `RawCbor`, a wrapper over `Array[Byte]` that provides sensible `equals`, `hashCode`,
+`toString` and other utilities.
+
+```scala
+val rawCbor: RawCbor = CborOutput.writeRawCbor(Thing(42, "foo"))
+```
+
+`RawCbor` can also be used as a field type in a case class in order to represent arbitrary CBOR-serialized data:
+
+```scala
+case class RawCborThing(id: Int, data: RawCbor)
+object RawCborThing extends HasGenCodec[RawCborThing]
+```
+
+## Explicitly sized collections
+
+CBOR lists and maps can be encoded in two ways:
+
+* with explicit number of elements at the beginning (more compact, allows preallocation of buffers and better
+  performance)
+* with unspecified number of elements at the beginning along with a _break_ byte (`0xFF`) at the end (better for streaming
+  applications)
+
+When serializing collections into CBOR, you can control whether they are explicitly sized. This is done with
+`SizePolicy` that can be passed into `CborOutput`, e.g.
+
+```scala
+val cborList = CborOutput.write(List(1, 2, 3), sizePolicy = SizePolicy.Required)
+```
+
+`SizePolicy` has three values:
+
+* `Required` - Collections are always written with explicit size in their CBOR representation. This may be slow for
+  collections that don't have a fast `.size` method, e.g. `List`
+* `Optional` - Only collections that have a fast `.size` method are written with explicit size, e.g. `Vector`. This is
+  the default behavior.
+* `Ignored` - All collections are written without explicit size
+
+## Using CBOR keys of arbitrary type
+
+`GenCodec` has a strong assumption that all object/map keys are `String`s, like in JSON. One of the consequences is that
+serializing a `Map[K, V]` with `GenCodec` requires an instance of `GenKeyCodec` for `K` so that keys can be converted to
+strings.
+
+However, CBOR has no such limitations. Map keys can be arbitrary CBOR values. There are some additional tools
+in `GenCodec` and `CborInput/Output` that can leverage this.
+
+### `CborKeyCodec`
+
+`CborInput.read` and `CborOutput.write` accept a `CborKeyCodec` parameter that defines translation between string-typed
+keys and arbitrary CBOR-encoded keys. You can use this e.g. to give a numeric label to every textual field, reducing the
+size of resulting CBOR.
+
+```scala
+val keyCodec = new CborKeyCodec {
+  def writeFieldKey(fieldName: String, output: CborOutput): Unit = fieldName match {
+    case "intField" => output.writeInt(1)
+    case "strField" => output.writeInt(2)
+    case name => output.writeString(name)
+  }
```

Comment:
This is an alternative to using `@cborKey` annotation, right?
Or is it more flexible?

Reply:
`CborKeyCodec` passed directly to `CborOutput`/`CborInput` is global - it will affect all the keys from all case classes, maps, etc. This is good e.g. for encoding SenML which has numeric labels for fields.

`@cborKey` applies only to some particular field of some particular case class (or discriminator value of a case class/object).